infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntrySinkBufferManager.java

```diff
@@ -84,8 +84,7 @@ public void processBuffer() {
             LogReplicationEntryMetadataMsg metadata = entry.getMetadata();
             if (metadata.getTimestamp() <= lastProcessedSeq) {
                 buffer.remove(metadata.getPreviousTimestamp());
-            } else if (metadata.getPreviousTimestamp() <= lastProcessedSeq) {
-                sinkManager.processMessage(entry);
+            } else if (metadata.getPreviousTimestamp() <= lastProcessedSeq && sinkManager.processMessage(entry)) {
```

Comment:
1. I think the order should not be impacted since we are checking that the current msg has a previousTimestamp which was processed by Standby + new data...So the order would be maintained by the first part of the check and that would make the current msg to be in the expected sphere..
2. If we check for only equality, we would miss msgs when active's metadata is behind what standby has processed. (Like the example in the desc).

> Example: 
> Active: [1,4,5,7]               ->                 Standby [1,4,5,7] Replicated.
> **_previousTimestamp = -1.                       lastProcessedTimestamp = 7_**
> 
> Active: [8,10,17].            ->.                 Standby[8,10,17] Replicated
> (ACK miss, LR stops                             _**lastProcessedTimestamp = 17**_
> so active remains in the 
> previous state)
> **_previousTimestamp = -1_**
> 
> Acive: [8, 10, 17, 22]      ->      Standby should process this msg since "22" is a new entry, and order would be correct

3. In the processBuffer(), we remove the entries which are older than the lastProcessedTimestamp. And since the sequencer will never reassign a token, even if the the msg in the buffer has an old snapshot, that should be removed, right ?

> for (LogReplicationEntryMsg entry : buffer.values()) {
>             LogReplicationEntryMetadataMsg metadata = entry.getMetadata();
>             if (metadata.getTimestamp() <= lastProcessedSeq) {
>                 buffer.remove(metadata.getPreviousTimestamp());

Reply:
Yes, you're right I forgot this is a batch of messages and was thinking on individual messages (which is incorrect). So, yes <= is correct. 

Well it will stay there if the new deltas (lastProcessedSeq) don't progress beyond the stale entry's timestamp (metadata.getTimestamp()), but eventually it will get removed. It should be expected that it always goes beyond as the snapshot base address can't regress, so yes agree we are fine as it will get removed at some point.